lib/FusionInventory/Agent/Task/Inventory/Linux/Storages/Megacli2.pm

```diff
@@ -0,0 +1,145 @@
+package FusionInventory::Agent::Task::Inventory::Linux::Storages::Megacli2;
+
+use strict;
+use warnings;
+
+use parent 'FusionInventory::Agent::Task::Inventory::Module';
+
+use File::Basename qw(basename);
+use FusionInventory::Agent::Tools;
+use FusionInventory::Agent::Tools::Linux qw(getInfoFromSmartctl);
+
+use constant RE => qr/^([^:]+?)\s*:\s*(.*\S)/;
+
+sub isEnabled {
+    return canRun('megacli') && canRun('smartctl');
+}
+
+sub doInventory {
+    my (%params) = @_;
+
+    my $inventory = $params{inventory};
+
+    my $count = getFirstMatch(
+        command => "megacli -adpCount",
+        pattern => qr/Controller Count: (\d+)/
+    );
+    return unless $count;
+
+    for (my $adp = 0; $adp < $count; $adp++) {
+        my $adpinfo = _getAdpPciInfo(adp => $adp);
+        my $block = _adpInfoToBlock($adpinfo);
+        my $pdlist = _getPDlist(adp => $adp);
+        my $ldinfo = _getLDinfo(adp => $adp);
+
+        while (my ($id, $pd) = each %$pdlist) {
+            # JBOD and Non-RAID are processed by the parent module, skip.
+            next if (
+                (defined $pd->{__diskgroup} &&
+                 defined $ldinfo->{ $pd->{__diskgroup} } &&
+                 defined $ldinfo->{ $pd->{__diskgroup} }->{'Number Of Drives'} &&
+                 defined $ldinfo->{ $pd->{__diskgroup} }->{'Name'} &&
+                 $ldinfo->{ $pd->{__diskgroup} }->{'Number Of Drives'} eq '1' &&
+                 $ldinfo->{ $pd->{__diskgroup} }->{'Name'} =~ /Non\s*\-*RAID/i)
+                    ||
+                $pd->{'Firmware state'} =~ /JBOD/i
+            );
+
+            my $storage = getInfoFromSmartctl(
+                device => '/dev/' . $block,
+                extra  => '-d megaraid,' . $id);
+
+            $inventory->addEntry(
+                section => 'STORAGES',
+                entry   => $storage,
+            );
+        }
+    }
+}
+
+sub _getPDlist {
+    my (%params) = @_;
+
+    $params{command} = defined $params{adp} ? "megacli -pdlist -a$params{adp}" : undef;
+
+    my (%srclist, %pdlist);
+    my $n = 0;
+
+    foreach my $line (getAllLines(%params)) {
+        next unless $line =~ RE;
+        my ($key, $val) = ($1, $2);
+        $n++ if $key =~ /Enclosure Device ID/;
+        $srclist{$n}->{$key} = $val;
+    }
+
+    for my $val (values %srclist) {
```

Comment:
Can you be consistent in the for loop syntax and use `foreach` here ? (unless you manage to merge the 2 loops ;-))

Reply:
[The foreach keyword is actually a synonym for the for keyword, so you can use either.](https://perldoc.perl.org/perlsyn#Foreach-Loops) ;) But i'll do.